### ISSUE TITLE
[bug]: 3.4.0 update causes issue with xformers & torch+cu121 reverting to torch+cpu

### ISSUE THREAD

**Issue by Millu** [invoke-ai/InvokeAI#5065]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### OS

Linux

### GPU

cuda

### What version did you experience this issue on?

3.4.0rc3

### What happened?

When updating to 3.4.0rc3, an issue occurred with `xformers` and the installed torch version: 

Error below:

```
WARNING[XFORMERS]: xFormers can't load C++/CUDA extensions. xFormers was built for:
    PyTorch 2.1.0+cu121 with CUDA 1201 (you have 2.1.0+cpu)
    Python  3.10.11 (you have 3.10.6)
  Please reinstall xformers (see https://github.com/facebookresearch/xformers#installing-xformers)
  Memory-efficient attention, SwiGLU, sparse and more won't be available.
  Set XFORMERS_MORE_DETAILS=1 for more details
```

InvokeAI will still run, but generations only run on the CPU, not the GPU

### Contact Details

@imic on Discord

**Comment by psychedelicious** [comment]
discord thread: https://discord.com/channels/1020123559063990373/1172065190905384981

There are a good number of users with this problem sprinkled throughout gh and discord

The reason for bumping to CUDA 12.1 was that xformers didn't have a build for 11.8. This has since been resolved, and now we do not need to enforce CUDA 12.1. see https://github.com/facebookresearch/xformers/issues/897

I think we can safely revert that change from 11.8 to 12.1, and advise users who were affected to reinstall.

---

I did a lot of troubleshooting on this today trying to reproduce the issue without success.

TL;DR: I think the fail-forward path may be for users to update their OS's CUDA stuff to 12.1.

I nuked all nvidia everything on my system - drivers, cuda, everything - and reinstalled, being careful to install only CUDA 11.8 stuff. Also created a fresh venv and installed everything from the 11.8 index. Confirmed my OS had only CUDA 11.8 and the invoke venv only had CUDA 11.8 torch/torchvision/xformers.

The app worked fine in this state.

Then I created a fresh venv, installing CUDA 12.1 everything. Still worked fine.

Maybe the trick is to *only* update xformers? Dunno, couldn't figure it out. Probably people just need to update their OS CUDA to 12.1, which makes sense to me.

**Comment by Qualzz** [comment]
is there a way to go back to the previous state without reinstalling everything ?

**Comment by psychedelicious** [comment]
@Qualzz Thanks for sharing the fix that worked for you.

For less-technical users, reinstalling is probably the simplest solution.

Note that "reinstall" is only reinstalling the application's python packages - not user data like models or generations. All of that is unaffected by a reinstall.

**Comment by psychedelicious** [comment]
To stick with the CUDA 11.8 torch builds, you may be able to repair an existing installation like so:
- Run the launcher
- Select the dev console option
- Run `pip install ".[xformers]" --upgrade --force-reinstall --extra-index-url https://download.pytorch.org/whl/cu118` (may take some time to download torch and install)
- Run the app again

**Comment by psychedelicious** [comment]
I did a lot of testing today on a windows machine. 

This was complicated by torch having version 2.1.1 on pypi. It's not a proper release, but is present in the indices. Our torch dep is defined as `torch~=2.1.0`, and for whatever silly reason, if you use `--extra-index-url https://download.pytorch.org/whl/cu121`, you end up with torch 2.1.1 cu121. 

Of course, this breaks things (and was very confusing).

Scenarios:
1. Install 3.3.0post3, then update to 3.4.0rc4
    - Issue occurs ❌ 
    - You get torch 2.1.0 w/ CPU ❌ 
2. Install 3.3.0post3, then install 3.4.0rc4 on top 
    - Issue doesn't occur ✅ 
    - You get torch 2.1.0 w/ CUDA 11.8 ❌
3. Install 3.4.0rc4 directly, no existing install
    - Issue doesn't occur ✅ 
    - You get torch 2.1.0 w/ CUDA 11.8 ❌
4. Install 3.3.0post3, then update to 3.4.0rc4, targeting a branch with `torch==2.1.0`
    - Issue doesn't occur ✅ 
    - You get torch 2.1.0 w/ CUDA 11.8 ❌
5. Install 3.3.0.post3, add `--extra-index-url https://download.pytorch.org/whl/cu121` to the `pip install` in `invokeai_update.py`, then update to 3.4.0rc4, targeting a branch with `torch==2.1.0`
    - Issue doesn't occur ✅ 
    - You get CUDA 12.1 ✅ 

Action items:
1. `torch` should never use `~=` to specify its version and should use `torch==2.1.0` for now
2. If we want to update the CUDA version, the update must provide the extra index URL so torch can be installed with the correct CUDA version, else it may fall back to CPU.
    
    This modification to `invokeai_update.py` works: https://github.com/invoke-ai/InvokeAI/commit/4c37c48b8c36d5defd6348a2e1c770d2e68f179f
    
    But it hardcodes the index and is tricky to deploy, because we cannot update an existing install's update script. Not sure how best to do this

**Comment by Millu** [comment]
For anyone who is experiencing this issue, launch your invoke.bat/invoke.sh, select the option to open the developer console and run:

`pip install ".[xformers]" --upgrade --force-reinstall --extra-index-url https://download.pytorch.org/whl/cu121`

If you run into an error with `typing_extension`, re-open the developer console and run: 

`pip install -U typing-extensions`

**Comment by makemefeelgr8** [comment]
> 2\. Run:
>     `pip install ".[xformers]" --upgrade --force-reinstall --extra-index-url https://download.pytorch.org/whl/cu121`

The command throws on Windows:

```log
ERROR: Directory '.[xformers]' is not installable. Neither 'setup.py' nor 'pyproject.toml' found.
```

I had to change it like this to make it work:

```cmd
pip install xformers --upgrade --force-reinstall --extra-index-url https://download.pytorch.org/whl/cu121
```

It also threw the following error in the end:
```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
datasets 2.14.6 requires fsspec[http]<=2023.10.0,>=2023.1.0, but you have fsspec 2023.12.2 which is incompatible.
```

What's interesting, image generation speed actually increased, so the update was worth it after all.

**Comment by Eugeniusz-Gienek** [comment]
For me the full change which worked is as follows:
`pip install --force-reinstall torch==2.1.2+cu121 torchvision==0.16.2+cu121 --extra-index-url https://download.pytorch.org/whl/cu121 xformers==0.0.23.post1 tensorboard==2.15.1 tensorflow==2.15.0.post1
pip install -U open-clip-torch blendmodes mediapipe`

**Comment by psychedelicious** [comment]
The installer now does a force install to mitigate this kind of problem.

**Comment by Cristy94** [comment]
```
pip install --force-reinstall torch==2.1.2+cu121 torchvision==0.16.2+cu121 --extra-index-url https://download.pytorch.org/whl/cu121 xformers==0.0.23.post1 tensorboard==2.15.1 tensorflow==2.15.0
```

This worked for me
